test/object-store/sync/flx_role_change.cpp

```diff
@@ -857,4 +857,298 @@ TEST_CASE("flx: role changes during bootstrap complete successfully", "[sync][fl
+          "[sync][flx][baas][role change][client reset]") {
+    auto logger = util::Logger::get_default_logger();
+    static std::unique_ptr<FLXSyncTestHarness> harness;
+
+    // 150 emps, 25 mgrs, 10 dirs
+    // 10 objects before flush
+    // 512 download soft max bytes
+    TestParams params{};
+    params.max_download_bytes = 512;
+    // size_t total_num = params.num_emps + params.num_mgrs + params.num_dirs;
+    if (!harness) {
+        harness = setup_harness("flx_role_change_during_cr", params);
+    }
+    REQUIRE(harness);
+
+    SECTION("Role change during client reset") {
+        // Get the current rules so it can be updated during the test
+        auto& app_session = harness->session().app_session();
+        auto default_rule = app_session.admin_api.get_default_rule(app_session.server_app_id);
+
+        enum ClientResetTestState {
+            not_ready,
+            start,
+            bind_before_cr_session,
+            cr_session_ident,
+            cr_session_downloading,
+            cr_session_downloaded,
+            cr_session_integrating,
+            cr_session_integrated,
+            bind_after_cr_session,
+            merged_after_cr_session,
+            ident_after_cr_session,
+        };
+
+        // int update_msg_count = -1;
+        // int bootstrap_count = 0;
+        bool client_reset_error = false;
+        bool role_change_error = false;
+        std::optional<ClientResetTestState> update_role_state = std::nullopt;
+        int client_reset_count = 0;
+        TestingStateMachine<ClientResetTestState> client_reset_state(ClientResetTestState::not_ready);
+
+        auto setup_config_callbacks = [&](SyncTestFile& config) {
+            // Use the sync client event hook to check for the error received and for tracking
+            // download messages and bootstraps
+            config.sync_config->on_sync_client_event_hook = [&](std::weak_ptr<SyncSession> session_ptr,
+                                                                const SyncClientHookData& data) {
+                bool is_fresh_path;
+                if (auto session = session_ptr.lock()) {
+                    is_fresh_path = _impl::client_reset::is_fresh_path(session->path());
+                }
+                else {
+                    //  Session is not valid anymore... exit now
+                    return SyncClientHookAction::NoAction;
+                }
+
+                client_reset_state.transition_with(
+                    [&](ClientResetTestState cur_state) -> std::optional<ClientResetTestState> {
+                        using BatchState = sync::DownloadBatchState;
+                        using Event = SyncClientHookEvent;
+
+                        if (cur_state == ClientResetTestState::not_ready)
+                            return std::nullopt;
+
+                        // Record that the error occurred
+                        if (data.event == Event::ErrorMessageReceived) {
```

Comment:
why don't you handle the errors in the switch below?

Reply:
Because there is an early return if the `update_role_state` is empty, which means the role change has been performed and the test doesn't care about the current state anymore. The errors are handled above this since the 200 error will come after the role change has been triggered.
I'll add an early return in this `if` block, since it doesn't need to continue if an error is received.